web/Services/PostcodeService.cs

```diff
@@ -0,0 +1,176 @@
+        }
+        else
+        {
+            postcode = "";
+        }
+
+        return postcode;
+    }
+
+    private void AddPostcodeToPostcodeDb(PostcodeioResponseModel postcodeResponse)
+    {
+        if (postcodeResponse.Result == null || postcodeResponse.Result.Lat == null || postcodeResponse.Result.Lon == null || postcodeResponse.Result.Postcode == null)
+        {
+            _logger.LogWarning("Postcode cannot be converted to coordinates.");
+        }
+        else
+        {
+            var postcodeDbModel = new PostcodeDbModel
+            {
+                Postcode = postcodeResponse.Result.Postcode,
+                Lat = postcodeResponse.Result.Lat,
+                Lon = postcodeResponse.Result.Lon,
+            };
+            _dbContext.Postcodes.Add(postcodeDbModel);
+        }
+        _dbContext.SaveChanges();
+    }
+    
+    public async Task AddSinglePostcodeToDatabaseIfAbsent(string postcode)
+    {
+        if (postcode != "" && !_dbContext.Postcodes.Any(p => p.Postcode == postcode))
+        {
+            var responseBody = await MakeSingleApiRequestToPostcodeApi(postcode);
+            var postcodeResponse = TurnResponseBodyToModel(responseBody);
+            AddPostcodeToPostcodeDb(postcodeResponse);
+        }
+    }
+
+    private async Task<string> MakeSingleApiRequestToPostcodeApi(string postcode)
+    {
+        var escapedPostalCode = Uri.EscapeDataString(postcode);
+        var uri = $"https://api.postcodes.io/postcodes/{escapedPostalCode}";
+        var responseBody = string.Empty;
+        _logger.LogInformation("Making API Request to {Uri}", uri);
+        try
+        {
+            var response = await _client.GetAsync(uri);
+            response.EnsureSuccessStatusCode();
+            responseBody = await response.Content.ReadAsStringAsync();
+            _logger.LogInformation("Successful API request to {Uri}", uri);
+        }
+        catch (Exception e)
+        {
+            _logger.LogWarning("Message :{Message} ", e.Message);
+        }
+
+        return responseBody;
+    }
+    
+    private PostcodeioResponseModel TurnResponseBodyToModel(string responseBody)
+    {
+        if (string.IsNullOrEmpty(responseBody))
+        {
+            return new PostcodeioResponseModel();
+        }
+        var postcodeResponse = JsonConvert.DeserializeObject<PostcodeioResponseModel>(responseBody);
+        return postcodeResponse!;
+    }
+
+    public async Task BulkAddPostcodesToDatabaseIfAbsent(List<string> postcodes)
+    {
+        var postcodesToLookup = new List<string>();
+        foreach (string postcode in postcodes)
+        {
+            if (!_dbContext.Postcodes.Any(p => p.Postcode == postcode) && !postcodesToLookup.Any(p => p == postcode))
+            {
+                postcodesToLookup.Add(postcode);
+            }
+
+            if (postcodesToLookup.Count() == 100)
+            {
+                await BulkGetLatLonForPostcodes(postcodesToLookup);

```

Comment:
instead of awaiting these each time, we can make a list of tasks and add these onto it, only awaiting all of them together at the end, that would prevent any blocking :)

Reply:
Would the state of postcodesToLookup be saved at the moment of the Task being added to the list, or would the fact that it immediately gets set to empty cause any problems storing a list of Tasks to await later?